### ISSUE TITLE
ICE when undeclared symbol in struct initializer

### ISSUE THREAD

**Issue by chloekek** [ispc/ispc#2509]
### ISPC version

Intel(r) Implicit SPMD Program Compiler (Intel(r) ISPC), 1.19.0 (build commit 147cc795073f50b6 @ 20230227, LLVM 14.0.6)

### Reproducible example

<https://godbolt.org/z/rexKdxcja>

```ispc
struct S
{
    float x;
};

void f()
{
    S s = { b };
}
```

### Compiler output

```
<source>:8:13: Error: Undeclared symbol "b". Did you mean "f"? 
    S s = { b };
            ^

ispc: /usr/local/src/llvm/bin-14.0/include/llvm/Support/Casting.h:104: static bool llvm::isa_impl_cl<ispc::ExprList, const ispc::Expr *>::doit(const From *) [To = ispc::ExprList, From = const ispc::Expr *]: Assertion `Val && "isa<> used on a null pointer"' failed.
/usr/local/src/ispc/src/main.cpp(412): FATAL ERROR: Unhandled signal sent to process; terminating.
***
*** Please file a bug report at https://github.com/ispc/ispc/issues
*** (Including as much information as you can about how to reproduce this error).
*** You have apparently encountered a bug in the compiler that we'd like to fix!
***
Compiler returned: 139
```

**Comment by nurmukhametov** [comment]
> At least, it has the message about undeclared symbol.
> 
> There is probably the variation of same problem which is more confusing:
> 
> ```c
> struct D {
>     int h;
> };
> struct D d;
> int foo(varying int a) {
>      return a += d.f;
> }
> ```
> 
> ```shell
> /usr/local/src/ispc/src/main.cpp(412): FATAL ERROR: Unhandled signal sent to process; terminating.
> ***
> *** Please file a bug report at https://github.com/ispc/ispc/issues
> *** (Including as much information as you can about how to reproduce this error).
> *** You have apparently encountered a bug in the compiler that we'd like to fix!
> ***
> Compiler returned: 139
> ```
> 
> Compiler explorer link: https://godbolt.org/z/csr4qfWcP

It is a different issue, see #2514

**Comment by mingodad** [comment]
Looking through the stack trace while debugging the change shown bellow seems to fix this issue:
```
--------------------------------- src/stmt.cpp ---------------------------------
index 1fb8094a..7db49716 100644
@@ -269,7 +269,7 @@ static bool checkInit(const Type *type, Expr **init) {
         // If it's an expr list with an atomic type, we'll later issue
         // an error.  Need to leave vars[i].init as is in that case so
         // it is in fact caught later, though.
-        if (llvm::dyn_cast<ExprList>(*init) == nullptr) {
+        if (*init != nullptr && llvm::dyn_cast<ExprList>(*init) == nullptr) {
             const Type *t = (*init) ? (*init)->GetType() : nullptr;
             if (t && t->IsDependentType()) {
                 return false;
```

**Comment by nurmukhametov** [comment]
> Looking through the stack trace while debugging the change shown bellow seems to fix this issue:
> 
> ```
> --------------------------------- src/stmt.cpp ---------------------------------
> index 1fb8094a..7db49716 100644
> @@ -269,7 +269,7 @@ static bool checkInit(const Type *type, Expr **init) {
>          // If it's an expr list with an atomic type, we'll later issue
>          // an error.  Need to leave vars[i].init as is in that case so
>          // it is in fact caught later, though.
> -        if (llvm::dyn_cast<ExprList>(*init) == nullptr) {
> +        if (*init != nullptr && llvm::dyn_cast<ExprList>(*init) == nullptr) {
>              const Type *t = (*init) ? (*init)->GetType() : nullptr;
>              if (t && t->IsDependentType()) {
>                  return false;
> ```

Thank you. I have opened PR with suggested fix #2548.

**Comment by nurmukhametov** [comment]
Fixed by #2548
